### PR TITLE
Add clustering interface

### PR DESCRIPTION
**Pull request by edwintorok**
This adds a new `xcp.cluster` library for the new clustering interface and doesn't touch any of the existing interfaces:
```
git diff origin/master --stat
 cluster/cluster_client.ml    |  14 ++++++++++
 cluster/cluster_idl.ml       |   3 +++
 cluster/cluster_interface.ml | 214 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 cluster/jbuild               |  40 ++++++++++++++++++++++++++++
 4 files changed, 271 insertions(+)
```

The spec file doesn't require any changes.

The purpose of this PR is to reduce the diff against master to make merging `feature/REQ477` easier.

The commit history here reflects the development, could consider squashing them if preferred.

### CODE REVIEW

File: cluster/cluster_interface.ml [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+type debug_info = string
+[@@doc ["An uninterpreted string associated with the operation."]]
+[@@deriving rpcty]
+
+type cluster_name = string
+[@@doc ["Name of the cluster"]]
+[@@deriving rpcty]
+
+type address = IPv4 of string
+[@doc ["An IP address"]]
```
Comment (lindig) [inline cluster/cluster_interface.ml]: Should this say "An IP address a.b.c.d" or similar to indicate that this is IPv4 and not a hostname or IPv6?

File: cluster/cluster_interface.ml [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+        "There is no authentication on these, but they are only available on the local machine.";
+      ];
+      version = (1,0,0);
+    }
+
+  let implementation = implement description
+
+  let create = declare
+      "create"
+      ["Creates the cluster. The call takes the initial config of";
```
Comment (lindig) [inline cluster/cluster_interface.ml]: When we concatenate these strings, we get something like ".. of;the ..". Is this intended?

Comment (edwintorok) [reply cluster/cluster_interface.ml]: The `;` won't be part of the concatenation, it separates the strings in the ocaml list.

Comment (lindig) [reply cluster/cluster_interface.ml]: That's right. So the concatenation would add a space?

Comment (edwintorok) [reply cluster/cluster_interface.ml]: Yes it concatenates the strings with spaces: https://github.com/mirage/ocaml-rpc/blob/474d9cd1a10de4f6078a6422e14a91ed84d8173b/lib/markdowngen.ml#L97

**Comment by mseri** [comment]
Can this be merged?

File: cluster/cluster_interface.ml [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+
+open Idl
+
+let service_name = "cluster"
+let queue_name = Xcp_service.common_prefix ^ service_name
+let json_path = "/var/xapi/cluster.json"
+let xml_path = "/var/xapi/cluster"
+
+type debug_info = string
+[@@doc ["An uninterpreted string associated with the operation."]]
```
Comment (mseri) [inline cluster/cluster_interface.ml]: Just for reference: now you can use actual docstrings here, and they will work as well

File: cluster/jbuild [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+let () = Printf.ksprintf Jbuild_plugin.V1.send {|
+(jbuild_version 1)
+
+(library
+ ((name xcp_cluster)
+  (public_name xcp.cluster)
+  (modules (:standard))
+  (flags (:standard -w -39 %s))
+  (libraries (xcp threads rpclib))
+  (wrapped false)
```
Comment (mseri) [inline cluster/jbuild]: We should start wrapping `xcp-idl` libraries

**Comment by lindig** [comment]
Do we want the history to be cleaned up? This is new code - so in my opinion this could be also squashed or we could have one commit per file.

File: cluster/jbuild (hunk 4 above)
Comment (mseri) [reply cluster/jbuild]: Please add the executable rules for `cluster_client` and call it `cluster_cli` in line with the others. Do not add a `public_name` or an alias. If https://github.com/xapi-project/xcp-idl/pull/198 is merged before this PR, please update the README

File: cluster/jbuild [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+  if is_coverage then "bisect_ppx" :: rewriters_ppx else rewriters_ppx
+
+
+let () = Printf.ksprintf Jbuild_plugin.V1.send {|
+(jbuild_version 1)
+
+(library
+ ((name xcp_cluster)
+  (public_name xcp.cluster)
+  (modules (:standard))
```
Comment (mseri) [inline cluster/jbuild]: I think you should do `(modules (:standard \ cluster_client))` and completely delete `cluster_idl`

File: cluster/cluster_client.ml [hunk 6]
```diff
@@ -0,0 +1,14 @@
+
+open Xcp_client
```
Comment (mseri) [inline cluster/cluster_client.ml]: In line with comments in new clients, please don't open this globally, but add `let open Xcp_client` inside the `rpc` function body

File: cluster/jbuild (hunk 5 above)
Comment (mseri) [reply cluster/jbuild]: Sorry, an oversight. I meant to add a `cluster_cli` as with the other components. Then remove that from the `modules` list, and add the executable rules

File: cluster/jbuild (hunk 4 above)
Comment (edwintorok) [reply cluster/jbuild]: Ok, I'll update the PR, you are right we can drop cluster_idl.ml in that case.

**Comment by edwintorok** [comment]
Pushed some cleanups, I agree that we should squash it all at merge time.
Note that I changed the API of the `rpc` function, it used  to take a function for building the URL, the cluster daemon  / xapi integration doesn't support anything other than message-switch at the moment, so I just made it similar to the other interfaces which take a url directly.

**Comment by edwintorok** [comment]
The failure is the same on as in https://github.com/xapi-project/xcp-idl/pull/198 due to coverage script

**Comment by edwintorok** [comment]
Isn't it a bit redundant to have `Xcp_cluster.Cluster_interface`? Thought that `Xcp_cluster.Interface` would be better now that I dropped the wrapping.

**Comment by mseri** [comment]
It does, but I'd rather keep things consistent and we cannot change all the other interfaces for the moment

**Pull request merged**